test/integration/socket_interface_integration_test.cc

```diff
@@ -0,0 +1,43 @@
+#include "common/network/socket_interface.h"
+
+#include "test/integration/integration.h"
+#include "test/test_common/environment.h"
+#include "test/test_common/utility.h"
+
+#include "gtest/gtest.h"
+
+namespace Envoy {
+namespace {
+
+class SocketInterfaceIntegrationTest : public BaseIntegrationTest,
+                                       public testing::TestWithParam<Network::Address::IpVersion> {
+public:
+  SocketInterfaceIntegrationTest() : BaseIntegrationTest(GetParam(), config()) {
+    use_lds_ = false;
+  };
+
+  static std::string config() {
+    // At least one empty filter chain needs to be specified.
+    return absl::StrCat(ConfigHelper::httpProxyConfig(), R"EOF(
+bootstrap_extensions:
+  - name: envoy.extensions.network.socket_interface.default_socket_interface
+    typed_config:
+      "@type": type.googleapis.com/envoy.extensions.network.socket_interface.v3.DefaultSocketInterface
+default_socket_interface: "envoy.extensions.network.socket_interface.default_socket_interface"
+    )EOF");
+  }
+};
+
+INSTANTIATE_TEST_SUITE_P(IpVersions, SocketInterfaceIntegrationTest,
+                         testing::ValuesIn(TestEnvironment::getIpVersionsForTest()),
+                         TestUtility::ipTestParamsToString);
+
+TEST_P(SocketInterfaceIntegrationTest, Basic) {
+  BaseIntegrationTest::initialize();
+  const Network::SocketInterface* factory = Network::socketInterface(
```

Comment:
Wouldn't this be the case even if the registration didn't work? I think what I'm proposing is to define a shim extension in this integration test, and just internally have it return the default implementation, then run some very basic echo like integration test.

Reply:
At this point no. We have a static global initialization of the default `SocketInterface` done via `ScopedInjectableLoader`, a `ScopedInjectableLoader`. In this test, that is overridden by the `config()` above which initializes the default interface as a bootstrap extension and uses it to reinitialize the `SocketInterfaceSingleton` (done in the server init phase). We did discuss with @lizan the option to completely remove the static global initialization, but we'll have to update some tests to support that, as not all of them initialize the server. So that's now on the ever growing todo list :-)

As for the test, this does start a listener through the bootstrap initialized `SocketInterface`. I guess we could add something like the Hello echo test from `echo_integration_test.cc`, or were you thinking about something more complex?

Reply:
OK fair enough. Yeah it would be good to get rid of the scoped singleton if possible but agreed that can be done later.

> I guess we could add something like the Hello echo test from echo_integration_test.cc, or were you thinking about something more complex?

Yes exactly this would be fine.

Reply:
Perfect! Done!